crucible-mir-comp/src/Mir/Compositional/Convert.hs

```diff
@@ -306,6 +308,86 @@ exprToTerm sym sc w4VarMapRef val = liftIO $ do
     term <- SAW.scVariable sc ec
     return term
 
+
+-- | Try to convert a Crucible register value into a SAW core `Term`, using
+-- a `CryTermAdaptor` to validate and convert references to slices.
+-- We check that references to slices match the expected length specified
+-- in the `CryTermAdaptor`, and if so we read out the value stored in
+-- in the slice.
+regToTermWithAdapt :: forall m p sym t fs tp0 rtp args ret.
+    ( IsSymInterface sym,
+      sym ~ MirSym t fs,
+      m ~ OverrideSim (p sym) sym MIR rtp args ret) =>
+    sym ->
+    SAW.SharedContext ->
+    String ->
+    IORef (Map SAW.VarIndex (Some (W4.Expr t))) ->
+    CryTermAdaptor ->
+    TypeShape tp0 ->
+    RegValue sym tp0 ->
+    m SAW.Term
+regToTermWithAdapt sym sc name w4VarMapRef ada0 shp0 rv0 = go ada0 shp0 rv0
+  where
+    go :: forall tp.
+        CryTermAdaptor ->
+        TypeShape tp ->
+        RegValue sym tp ->
+        m SAW.Term
+    go ada shp rv =
+      case (ada, shp, rv) of
+        (NoAdapt, _, _) -> regToTerm sym sc name w4VarMapRef shp rv
+        (AdaptTuple as, TupleShape _ elems, ag) -> do
+            terms <- accessMirAggregate' sym elems as ag $ \_off _sz shp' rv' a -> go a shp' rv'
+            liftIO $ SAW.scTuple sc terms
+        (AdaptArray a, ArrayShape _ _ shp', vec) -> do
+            terms <- goVector a shp' vec
+            tyTerm <- shapeToTerm' sc a shp'
+            liftIO $ SAW.scVector sc tyTerm terms
+        (AdaptDerefSlice n, SliceShape _ty elT M.Immut tpr, Ctx.Empty Ctx.:> RV mirPtr Ctx.:> RV lenExpr) ->
+          case BV.asUnsigned <$> W4.asBV lenExpr of
+            Nothing ->
+              fail "Slice length is not statically known"
+
+            Just n1
+              | AsBaseType baseT <- asBaseType tpr ->
```

Comment:
What is the reason for requiring a Crucible base type here? As far as I can tell, the code below ought to work with any `TypeShape`, not just base types.

Reply:
I think in principle we could generalize this, but we'll have to see how to do it.  The immediate reason is that we don't have a `TypeShape` for the type of elements, and requiring a base type is easy to explain and to convert to a `TypeShape`.   I don't think the code, strictly speaking, requires having a `TypeShape` but without it we'd have to duplicate a bunch of stuff.    At present we don't store a `TypeShape` in these situations to avoid potentially infinite terms (as recursion may occur under references).

So I think for the time being let's leave it like that, and we could think on how to generalize if the use case arises.